### ISSUE TITLE
incorrect spec for `String::len`?

### ISSUE THREAD

**Issue by arnaudgolfouse** [creusot-rs/creusot#1164]
```rust
impl ShallowModel for String {
    type ShallowModelTy = Seq<char>;
}

#[ensures(result@ == self@.len())]
fn len(&self) -> usize;
```
The length of the string is in bytes, so this spec is not correct I think? Same for `str::split_at`. Although I guess there is no way no observe this at the moment...

**Comment by xldenis** [comment]
Hmm yea I guess we should relax that to >=?

**Comment by arnaudgolfouse** [comment]
This would work, but shouldn't we actually give the correct length? With a logical function `utf8_len(char) -> usize`
I guess this would make `str::len` hard on provers, but if we use `>=` it would be a bit useless.

**Comment by xldenis** [comment]
Depends on what we need it for really.

**Comment by jhjourdan** [comment]
Do we want to formalized Unicode? No.
So the only alternative we have is not supporting `len` at all, which is definitely an option.

**Comment by arpj-rebola** [comment]
>  So the only alternative we have is not supporting `len` at all, which is definitely an option.

Not really, `String::len` has interesting properties that can be relevant (e.g. the length of string concatenation is the sum of lengths of concatenated strings; and similarly for splitting and substrings). It just doesn't correspond to the notion of length in `Seq`. One possible way there might be formalizing some (very basic) contract of `str::as_bytes`, which behaves nicely and compositionally, and then define `str::len` based on that.

Is `Seq<char>` a good model for `String`, as opposed to `Seq<u8>`, though?

**Comment by Lysxia** [comment]
> Is Seq<char> a good model for String, as opposed to Seq<u8>, though?

I think `Seq<char>` lets us hide encoding details better. One can still switch to a `Seq<u8>`-based representation via `str::as_bytes`.
